### PR TITLE
[WIP] PyTorch Tutorial

### PR DESCRIPTION
**Pull request by MaximumEntropy**
Initial commit for the PyTorch tutorial with basic tensor ops & a little bit of autograd.

### CODE REVIEW

**Comment by ReyhaneAskari** [comment]
Hi Sandeep, thanks for the nice work. I have just added several comments inside the code blocks. I think it's a very good tutorial but you might not have enough time to cover all of it. I suggest to have a backup plan for that (mark some of the contents that you think are not very important and just skip them in case you were short in time).  For example the first 3 notebooks are important but the last two can be just passed quickly.
A quick question, how are you going to run on cuda during the tutorial?

**Comment by ReyhaneAskari** [comment]
I just talked with Pascal, he pointed to [this link ](https://sites.google.com/a/lisa.iro.umontreal.ca/lisainfo/getting-started/welcome/tutorials-kick-off-meeting/tutorials-2015/instructions-to-run-the-notebook-remotely) where it is showed how to run a notebook remotely. This way you will be able to run on one of the lab machines and use the GPUs. It is best to avoid explaining this part to the students as we don't have much time.

File: pytorch/1. The Torch Tensor Library and Basic Operations.ipynb [hunk 1]
```diff
@@ -0,0 +1,585 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## PyTorch Tutorial\n",
+    "MILA, November 2017"
```
Comment (lamblin) [inline pytorch/1. The Torch Tensor Library and Basic Operations.ipynb]: You can put your name here.
I've seen you have acknowledge sources in other notebooks, that's good :)

File: pytorch/1. The Torch Tensor Library and Basic Operations.ipynb [hunk 2]
```diff
@@ -0,0 +1,585 @@
...
+     "output_type": "stream",
+     "text": [
+      "torch.Size([2, 3])\n",
+      "torch.Size([2, 3, 4])\n"
+     ]
+    }
+   ],
+   "source": [
+    "y = torch.randn(2, 3, 4)\n",
+    "print(y[:, 1, :3].size())\n",
```
Comment (lamblin) [inline pytorch/1. The Torch Tensor Library and Basic Operations.ipynb]: That's not advanced indexing, though.
The simplest kind of advanced indexing is indexing by a tensor, for instance `print(y[[1, 0, 1, 1]])` (shape would be (4, 3, 4)).
More complex can be something like `print(y[[0, 1], [[1, 0], [2, 1]], [[3], [1]]])`, where the indices are broadcasted together and iterated jointly on (the shape here would be (2, 2), corresponding to indices (0, 1, 3), (1, 0, 3), (0, 2, 1), (1, 1, 1)).

File: pytorch/2. Autograd.ipynb [hunk 3]
```diff
@@ -0,0 +1,988 @@
...
+    "print('do/db : \\n %s ' % (do_dinputs_2[2]))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## Gradient buffers: `.backward()` and `retain_graph=True`\n",
+    "\n",
+    "1. Calling `.backward()` clears gradient buffers and \"erases the tape\".\n",
```
Comment (lamblin) [inline pytorch/2. Autograd.ipynb]: I find it a bit confusing, since you say later that "Calling .backward() multiple times will accumulate gradients into .grad and NOT overwrite them". Is the "gradient buffer" something different from ".grad"?

File: pytorch/2. Autograd.ipynb [hunk 4]
```diff
@@ -0,0 +1,988 @@
...
+    "print('o.grad_fn : %s ' % (o.grad_fn))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "### `.detach()`\n",
+    "\n",
+    "1. It is possible to detach variables from the graph by calling `.deatch()`. \n",
```
Comment (lamblin) [inline pytorch/2. Autograd.ipynb]: ".detach()"

File: pytorch/2. Autograd.ipynb [hunk 5]
```diff
@@ -0,0 +1,988 @@
...
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {
+    "collapsed": true
+   },
+   "source": [
+    "## Gradients w.r.t intermediate variables in the graph\n",
+    "\n",
+    "1. By default, PyTorch all gradient computations w.r.t intermediate nodes in the graph are ad-hoc.\n",
```
Comment (lamblin) [inline pytorch/2. Autograd.ipynb:809]: I'm not sure what that means

File: pytorch/2. Autograd.ipynb [hunk 6]
```diff
@@ -0,0 +1,988 @@
...
+    "\n",
+    "2. This is in the intrest of saving memory.\n",
```
Comment (lamblin) [inline pytorch/2. Autograd.ipynb]: "interest"

File: pytorch/3. Introduction to the Torch Neural Network Library.ipynb [hunk 7]
```diff
@@ -0,0 +1,320 @@
...
+    "### torch.nn.init\n",
+    "\n",
+    "Provides a set of functions for standard weight initialization techniques\n",
+    "\n",
+    "`import torch.nn.init as init`\n",
+    "\n",
+    "1. Calculate the gain of a layer based on the activation function - `init.calculate_gain('sigmoid')`\n",
+    "2. Uniform init - `init.uniform(tensor, low, high)`\n",
+    "3. Xavier uniform - `init.xavier_uniform(tensor, gain=init.calculate_gain('sigmoid'))`\n",
+    "4. Xavier normal - `init.xavier_uniform(tensor, gain=init.calculate_gain('tanh'))`\n",
```
Comment (lamblin) [inline pytorch/3. Introduction to the Torch Neural Network Library.ipynb]: Do you mean "init.xavier_normal"?

File: pytorch/2. Autograd.ipynb (hunk 3 above)
Comment (lamblin) [reply pytorch/2. Autograd.ipynb]: Also, maybe introduce `zero_grad()` here?

**Review by lamblin — APPROVED** [review]
You may have to skim over the last examples, just present them at a high-level and let people come back there after the presentation at their own pace.

File: pytorch/2. Autograd.ipynb (hunk 3 above)
Comment (MaximumEntropy) [reply pytorch/2. Autograd.ipynb]: Ah yes what I really meant is clears the tape. I shouldn't use the word gradient buffer. Presenting `zero_grad()` will probably require me talking about `torch.optim` since I believe the only way to clear the `.grad` attribute of a variable that isn't associated with an optimizer is ` variable.grad.data.zero_()`

File: pytorch/2. Autograd.ipynb (hunk 5 above)
Comment (MaximumEntropy) [reply pytorch/2. Autograd.ipynb:809]: Ah what I meant is that PyTorch does not retain the gradients w.r.t intermediate nodes in the graph in order to save memory. It only retains the gradient of the loss w.r.t to the things that we actually care about (like the parameters of our model) and the intermediate gradients are "ad-hoc" since they are only used to compute gradient of the loss w.r.t the model parameters. I'll make this clearer :)

File: pytorch/3. Introduction to the Torch Neural Network Library.ipynb (hunk 7 above)
Comment (MaximumEntropy) [reply pytorch/3. Introduction to the Torch Neural Network Library.ipynb]: Yes. Oops

**Comment by MaximumEntropy** [comment]
@ReyhaneAskari thanks! I was able to set things up on my laptop following those instructions.
